SI-9369 Fix pattern matcher warnings for diamond shaped inheritance.

gbasler opened the pull request:
A previous optimization (d44a86f432a7f9ca250b014acdeab02ac9f2c304) for
pattern matcher exhaustivity checks used a smarter encoding to ensure
that the scrutinee can be equal to one child only.
However, in case of traits between the root and leave type, a child can
be of several types and these types should not be in a mutually exclusive
group.

I realized that my previous approach was just too complicated, a simple
solution that was suggested by @retronym is good enough.
I wouldn't expect users to write big class hierarchies with many traits in it -
usually there are a few traits and potentially many classes involved, so
simply putting each trait in a new group ensures that no class can be in the
same group as a base trait.

Review by @retronym

gbasler commented:
/rebuild

SethTisue commented:
/rebuild

SethTisue commented:
@gbasler is it possible here the timeout here indicates a real problem with the patch?

retronym commented on src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala:
```diff
@@ -143,18 +143,60 @@ trait TreeAndTypeAnalysis extends Debugging {
+                val children = enumerateChildren(tpe.typeSymbol)
+                val filtered = filterChildren(children)
+                wl ++= filtered
+                baseTypesForType += (tpe -> tpe.baseTypeSeq.toList.toSet)
```
Comparing `Type` with `==` and `hashCode` is almost always a bad idea.

Perhaps you can instead use the base class symbols, rather than the base types, for this book keeping.

retronym commented:
What if we introduced:

```
trait Simple2 extends Prop
```

This should not be mutually exclusive with `Simple1`, as a class could extend both of them. AFAICT it would fall into the same group in the scheme described above.

Should we instead draw a distinction between classes and traits? Traits are never mutually exclusive, and classes don't have the diamond problem to deal with.

gbasler committed:
SI-9369 Fix pattern matcher warnings for diamond shaped inheritance.

A previous optimization (d44a86f432a7f9ca250b014acdeab02ac9f2c304) for
pattern matcher exhaustivity checks used a smarter encoding to ensure
that the scrutinee can be equal to one child only.
However, in case of traits between the root and leave type, a child can
be of several types and these types should not be in a mutually exclusive
group. A simple solution (hat tip to retronym) is to just put traits
and classes into separate groups.

gbasler commented:
@SethTisue There was indeed a problem with the code. I've just force pushed a much simpler solution.

retronym commented:
@gbasler Does this patch improve https://issues.scala-lang.org/browse/SI-9398 or https://issues.scala-lang.org/browse/SI-9399, or some [other tickets against the pattern matcher](https://issues.scala-lang.org/issues/?jql=status%20%3D%20open%20AND%20component%20%3D%20%22Pattern%20Matcher%22%20AND%20project%20%3D%20SI%20AND%20updated%3E%3D-52w%20ORDER%20BY%20updated%20DESC)?

gbasler commented:
I'm a bit puzzled about https://issues.scala-lang.org/browse/SI-9398. I will debug that one. It does not have traits so I don't think it's related. 
https://issues.scala-lang.org/browse/SI-9399 etc have all the same cause: the pattern matcher identifies member variables of scrutinees only by their name, not by name and enclosing class name. So it thinks they all refer to the same variable.

gbasler commented:
I couldn't find a ticket that is fixed by that patch as well...

retronym commented:
LGTM

Second pair of eyes by @adriaanm, please.

adriaanm commented:
:eyes: say LGTM!  Thanks, @gbasler!
